chain/stateless_block.go

```diff
@@ -0,0 +1,150 @@
+
+type StatelessBlock struct {
+	Prnt   ids.ID `json:"parent"`
+	Tmstmp int64  `json:"timestamp"`
+	Hght   uint64 `json:"height"`
+
+	Txs []*Transaction `json:"txs"`
+
+	// StateRoot is the root of the post-execution state
+	// of [Prnt].
+	//
+	// This "deferred root" design allows for merklization
+	// to be done asynchronously instead of during [Build]
+	// or [Verify], which reduces the amount of time we are
+	// blocking the consensus engine from voting on the block,
+	// starting the verification of another block, etc.
+	StateRoot ids.ID `json:"stateRoot"`
+
+	bytes []byte
+	id    ids.ID
+}
+
+func NewStatelessBlock(
+	parentID ids.ID,
+	timestamp int64,
+	height uint64,
+	txs []*Transaction,
+	stateRoot ids.ID,
+) (*StatelessBlock, error) {
+	block := &StatelessBlock{
+		Prnt:      parentID,
+		Tmstmp:    timestamp,
+		Hght:      height,
+		Txs:       txs,
+		StateRoot: stateRoot,
+	}
+	blkBytes, err := block.Marshal()
+	if err != nil {
+		return nil, err
+	}
+	block.bytes = blkBytes
+	block.id = utils.ToID(blkBytes)
+	return block, nil
+}
+
+func (b *StatelessBlock) ID() ids.ID    { return b.id }
+func (b *StatelessBlock) Bytes() []byte { return b.bytes }
+func (b *StatelessBlock) Size() int     { return len(b.bytes) }
+
+func (b *StatelessBlock) String() string {
+	return fmt.Sprintf("(BlockID=%s, Height=%d, ParentRoot=%s, Size=%d)", b.id, b.Hght, b.Prnt, len(b.bytes))
+}
+
+func (b *StatelessBlock) Marshal() ([]byte, error) {
+	size := ids.IDLen + consts.Uint64Len + consts.Uint64Len +
+		consts.Uint64Len + window.WindowSliceSize +
+		consts.IntLen + codec.CummSize(b.Txs) +
+		ids.IDLen + consts.Uint64Len + consts.Uint64Len
+
+	p := codec.NewWriter(size, consts.NetworkSizeLimit)
+
+	p.PackID(b.Prnt)
+	p.PackInt64(b.Tmstmp)
+	p.PackUint64(b.Hght)
+
+	p.PackInt(uint32(len(b.Txs)))
+	for _, tx := range b.Txs {
+		if err := tx.Marshal(p); err != nil {
+			return nil, err
+		}
+	}
+
+	p.PackID(b.StateRoot)
+	bytes := p.Bytes()
+	if err := p.Err(); err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+func UnmarshalBlock(raw []byte, parser Parser) (*StatelessBlock, error) {
+	var (
+		p = codec.NewReader(raw, consts.NetworkSizeLimit)
+		b StatelessBlock
+	)
+
+	p.UnpackID(false, &b.Prnt)
+	b.Tmstmp = p.UnpackInt64(false)
+	b.Hght = p.UnpackUint64(false)
+
+	// Parse transactions
+	txCount := p.UnpackInt(false) // can produce empty blocks
+	actionCodec, authCodec := parser.ActionCodec(), parser.AuthCodec()
+	b.Txs = []*Transaction{} // don't preallocate all to avoid DoS
+	for i := uint32(0); i < txCount; i++ {
+		tx, err := UnmarshalTx(p, actionCodec, authCodec)
+		if err != nil {
+			return nil, err
+		}
+		b.Txs = append(b.Txs, tx)
+	}
+
+	p.UnpackID(false, &b.StateRoot)
+
+	// Ensure no leftover bytes
+	if !p.Empty() {
+		return nil, fmt.Errorf("%w: remaining=%d", ErrInvalidObject, len(raw)-p.Offset())
+	}
+	b.bytes = raw
+	b.id = utils.ToID(raw)
+	return &b, p.Err()
+}
+
+func NewGenesisBlock(root ids.ID) (*ExecutionBlock, error) {
```

Comment:
The genesis block is chain specific, so the eventual goal is for the chain type to handle this rather than the VM needing to know about the genesis, but relying on the chain for all other blocks.